### PR TITLE
Add hooks support

### LINKED ISSUE #46: Add hooks support
Hi,
I have a use case where adding hooks support to `@angular-extensions/elements` would be beneficial. This idea is somewhat similar to #30, so maybe they could be implemented at the same time.

### Problem
Let's take an application with several Angular elements packaged as web components. Naturally, they may have several common dependencies that we may want to load only once per page (for example core Angular packages or `ngrx`). In order to improve UX and time to first meaningful paint, we postpone loading these dependencies until the components are requested by the current view.

With the current API of `@angular-extensions/elements` I don't see a way this could be easily achieved.

### Proposal
Add hooks to `ElementConfig`:
```ts
{
  url: 'https://my-awesome-component.com',
  hooks: {
    beforeLoad: () => resolveDependencies('my-component')
  }
}
```
The library will execute `beforeLoad`  hook before inserting script with the component's JS bundle. If the hook returns a Promise (which is necessary in this particular case), the library will wait for this Promise to resolve before proceeding with the request. If the Promise fails, the corresponding error component is shown instead.

Do you think this API makes sense for `@angular-extensions/elements`?

### PR DESCRIPTION
**Pull request by MagicCurlyHair**
Add `beforeLoad` and `afterLoad` hooks to `ElementConfig`. This closes #46 

@tomastrajan please take a look at the implementation. If everything is good, I will add unit tests to this PR.

### CODE REVIEW

File: projects/elements/src/lib/lazy-elements/lazy-elements-loader.service.ts [hunk 1]
```diff
@@ -99,9 +117,23 @@ export class LazyElementsLoaderService {
...
+          this.handleHook(hooksConfig.afterLoad, tag)
+            .then(notifier.resolve)
+            .catch(notifier.reject);
+        } else {
+          notifier.resolve();
+        }
+      };
       script.onerror = notifier.reject;
-      document.body.appendChild(script);
+      if (hooksConfig?.beforeLoad) {
```
Comment (tomastrajan) [inline projects/elements/src/lib/lazy-elements/lazy-elements-loader.service.ts]: What happens if there is more than one hook? 

eg from global config and from element config ? 

Should be also possible to specify global hooks ? (in root config ? )

Comment (MagicCurlyHair) [reply projects/elements/src/lib/lazy-elements/lazy-elements-loader.service.ts]: Added support for root-level hooks. Service will try to run element-level hook if present and will fall back to root-level hook. I think that is the most flexible approach as that way hooks can be overridden for certain elements.

Comment (tomastrajan) [reply projects/elements/src/lib/lazy-elements/lazy-elements-loader.service.ts]: @MagicCurlyHair sounds great!

**Comment by tomastrajan** [comment]
@MagicCurlyHair could you please add content to the API section of the demo SPA and maybe even a small example in the `Advanced` section, even something simple as display alert as a hook.

Then we are ready to merge :)

**Comment by MagicCurlyHair** [comment]
@tomastrajan 
Finally finished it (hopefully :smile:). Added description to api section, example to advanced page and unit tests for service loader. I also did a small refactoring of service loader tests along the way, hope you don't mind.

**Comment by tomastrajan** [comment]
@MagicCurlyHair epic! Will have a look soon! Thank you!

**Comment by tomastrajan** [comment]
@MagicCurlyHair ok this is epic, thank you!

**Pull request merged**